Automated VCR re-recording

cben opened the pull request:
Something I've long wanted, a script for hands-off VCR re-recording :tada:

Can use minishift or (todo: untested) bring your own openshift setting OPENSHIFT_MASTER_HOST env var.

## Steps I took for new VCR recording:

1. Installed minishift v1.9.0, with manageiq plugin added according to https://github.com/ManageIQ/guides/blob/master/providers/openshift.md
2. I run
    ```
    minishift start --vm-driver virtualbox --openshift-version v3.6.1 --metrics --memory 5G
    ```
    (metrics needed to use same VCR in manageiq-providers-kubernetes, at least as that spec is currently written.  lots of RAM needed for metrics.  `--metrics` for v3.7.0 doesn't work yet.)
3. Then waited for metrics to come up, as confirmed by:
    ```
    oc get pods -n openshift-infra
    curl --insecure "https://$(oc get route -n openshift-infra hawkular-metrics --template '{{.spec.host}}')"
    ```
4. Run the new script:
    ```
    ./spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh
    ```
5. git commit, profit :-)

## Issues I didn't fix

- Caught difference in number of images between old and new refresh! :bug:
  We already had a not-fully-explained off by 1 there.
  I would like to merge this as pending test and debug later (will open issue).
- TODO: Object counts not yet reproducible between runs.
  However it's mostly stable
- TODO: API_TOKEN gets stored in the VCR (it was in previous cassette, too).
  This may be a problem if somebody passes a long-lived env.
  - Found https://relishapp.com/vcr/vcr/docs/configuration/filter-sensitive-data, didn't understand how to use yet.
- Sometimes project creation errors as it's still cleaning up from project deletion — despite the script waiting.  I'll try to report openshift bug, but just re-running this script helps.  I decided not to introduce retrying for this, would make script flow much more complicated.

cc @yaacov @enoodle @Ladas 

Need this gaprindashvili/yes for improving quotas test
https://bugzilla.redhat.com/show_bug.cgi?id=1504560

Ladas commented:
@cben looking e.g at ContainerReplicator.count failure

```ruby
(byebug) pp ContainerReplicator.pluck(:ems_ref, :name)
[["7b36741b-dd83-11e7-8319-169d0f7fea7d", "docker-registry-1"],
 ["7beb37af-dd83-11e7-8319-169d0f7fea7d", "router-1"],
 ["bf9b8058-dd84-11e7-8319-169d0f7fea7d", "my-replicationcontroller-2"],
 ["ea539a31-dd83-11e7-8319-169d0f7fea7d", "hawkular-cassandra-1"],
 ["e518dc66-dd83-11e7-8319-169d0f7fea7d", "hawkular-metrics"],
 ["e61f1832-dd83-11e7-8319-169d0f7fea7d", "heapster"]]
```

this looks correct, but spec things we deleted 2, so there should be 3 total. Is it possible the env. is either not deterministic, or there are more changes happening?

cben commented:
PTAL.  Still getting small count difference between recordings, but it's reasonably stable and I want to move on.  Added text files which will make "what changed between recordings" much easier to track with future changes...

- [ ] I still need to look into Ladislav's comment above.  I've since re-recorded multiple times, but I probably have that recording in `many-vcrs` branch.
- [ ] TODO: I'm not sure I have image example I promised Ladislav.
- [x] Added quotas with scopes to template.  May need more changes later but don't know which yet.
- [x] I again see image count difference between old and graph refresh.  Marked pending, I'm happy I have it caught, debugging and fixing are out of scope for this PR.

moolitayer commented on spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb:
```diff
@@ -140,52 +146,49 @@ def full_refresh_test
+        expect(ContainerProject.count).to eq(object_counts['ContainerProject'])
+        expect(ContainerProject.active.count).to eq(object_counts['ContainerProject'] - 1)
+
+        pending("why graph refresh DELETES 1 image from DB? why old refresh archives 2?")
```
:+1:

moolitayer commented on spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb:
```diff
@@ -6,15 +6,17 @@
-    token = 'theToken'
+    # env vars for easier VCR recording, see test_objects_record.sh
+    hostname = ENV["API_HOST"] || "host.example.com"
+    token = ENV["API_TOKEN"] || "theToken"
```
We have some automation around cluster creation and there these are named:
```
OPENSHIFT_MASTER_HOST
OPENSHIFT_MANAGEMENT_ADMIN_TOKEN
```
IT could prove useful to have the same names, up to you.

moolitayer replied:
https://github.com/container-mgmt/cm-ops-flow/tree/a455c59f7cc7d00a6e534de570de3dbe24a53e4b#needed-environment-variables

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh:
```diff
@@ -0,0 +1,109 @@
+#!/bin/bash
+
+set -e  # abort on error
+
```
Since this isn't using functions can we have section headers for readability? e.g

```
#
# Ensure API Server
#
```

cben replied:
Added an `echo`, I'm using these as both headings in code and headings in script output.

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh:
```diff
@@ -0,0 +1,109 @@
+    echo 'or have minishift in $PATH and already running, e.g.'
+    echo '    minishift addons enable manageiq'
+    echo '    minishift start --vm-driver virtualbox --openshift-version v3.7.0'
+    exit 1
```
you could add a link to this page explaining deployment options:
http://manageiq.org/docs/guides/providers/openshift

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/refresher_after_openshift_deletions.txt:
```diff
@@ -0,0 +1,39 @@
+VCR generated by spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh against 192.168.99.100
```
File name isn't clear to me as it's openshift objects and not refresh or manageiq related

cben replied:
You mean the .txt name, or test_objects_record.sh name?
.txt name is just mirroring .yml name.
All names half-thought, better names welcome :)

I wonder if script should go to tools/ instead of deep under spec/?

moolitayer replied:
.txt name
suggestion `/refresher_after_openshift_deletions.txt/post_deletion_objects.txt/`

> I wonder if script should go to tools/ instead of deep under spec/?

Will this need distribution e.g would we be interested in run the script not from a manageiq copy
(in some automation flow)?

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh:
```diff
@@ -0,0 +1,109 @@
+  if which minishift && minishift status | grep -i 'openshift:.*running'; then
+    export API_HOST="$(minishift ip)"
+    eval $(minishift oc-env --shell bash) # Ensure oc in PATH
+    oc login -u system:admin # With minishift, we know we can just do this
```
where do we login to a non minishift cluster?
or do we assume we are already logged in?
I'd say running on the openshift master is a good assumption (API_HOST=localhost)

cben replied:
for non-minishift BYO, script expects you to already have logged in.

Running on openshift master is inconvenient — you need manageiq checked out & working (at least test DB) to run this script, and script leaves updated files you'll want to git commit.
I should document how to `oc adm policy add-cluster-role-to-user cluster-admin you` for remote login.

I've never actually tested the script in BYO mode.  Let me try...

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh:
```diff
@@ -0,0 +1,109 @@
+
+VCR_DIR="$(dirname "$(realpath "$0")")"
+
+cd "$(realpath "$0" | sed 's:\(manageiq-providers-openshift/\).*$:\1:')" # repo root
```
cd "$(git rev-parse --show-toplevel)"

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh:
```diff
@@ -0,0 +1,109 @@
+  oc start-build my-build-config-$ind
+done
+
+while out="$(oc get build --all-namespaces)"; echo "$out"; [ "$(echo "$out" | egrep --count --word 'Complete|Failed|Error')" -ne 3 ]; do
```
Please upcase all the variables

moolitayer commented:
Looks great and taking us a long way.
This is meant to be run from a developer machine right(not on an automation server)?

cben commented:
> This is meant to be run from a developer machine right(not on an automation server)?

Right, however it's very close to what we'd need for an actually frequently run integration test, if we'll want that.

cben commented:
> This is meant to be run from a developer machine right(not on an automation server)?

Right, however it's very close to what we'd need for an actually frequently run integration test, if we'll want that.

cben commented:
Oops, I was pushing to different branch, you didn't see changes I was claiming, sorry :-(
Pushed now.
AFAICT, last PR state you saw was https://github.com/cben/manageiq-providers-openshift/compare/bf456cf...ab7707b ; can't give you a github link to see only new parts, but if you want you can locally:
```
git remote add cben https://github.com/cben/manageiq-providers-openshift
git fetch cben
git diff ab7707b..cben/vcr-script
```
I believe all comments addressed (see recent commits).

@miq-bot add-label gaprindashvili/yes
will need new vcr for improving quotas test

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh:
```diff
@@ -0,0 +1,111 @@
+  sleep 3
+done
+
+for ind in 0 1 2; do
```
Upcase var please

cben replied:
done

moolitayer commented on spec/vcr_cassettes/manageiq/providers/openshift/container_manager/test_objects_record.sh:
```diff
@@ -0,0 +1,111 @@
+
+rm -v "$VCR_DIR"/refresher_after_deletions.{yml,txt} || true
+describe_vcr > "$VCR_DIR"/refresher_after_deletions.txt
+env RECORD_VCR=after_deletions bundle exec rspec "$SPEC" || echo "^^ FAILURES ARE POSSIBLE, YOU'LL HAVE TO EDIT THE SPEC"
```
Are FAILURES expected due to the status issue above? something else?

cben replied:
Due to recordings not being 100% reproducible — it's quite stable now but still usually at least something is off by one.  "An ideal world is left as an exercise to the reader."

Additionally, if you switch Openshift versions, or change template deliberately, you'll get certain differences.

In any case, this script should not abort on test fail — it should let you get both recordings, and then you can edit & run spec again and again...

Better wording welcome.

cben commented:
ping @moolitayer anything else needed here?

moolitayer commented:
Awesome :muscle:

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 3731370ee732f07861b6d4598ff0befcbace3dbf
Author: Mooli Tayer <mtayer@redhat.com>
Date:   Sun Dec 31 19:56:57 2017 +0200

    Merge pull request #75 from cben/vcr-script
    
    Automated VCR re-recording
    (cherry picked from commit 6e8842e4592433ea2ce7ade2fb2d8806da9b8053)
```
